### PR TITLE
[PROF-10656] Add faulting_address to crashtracker reports

### PR DESCRIPTION
**Pull request by ivoanjo**
# What does this PR do?

While investigating a customer crash, I found that I dearly missed having the address that caused a segfault.

E.g. given this code:

```c
  char *potato = 0x1234;
  fprintf(stderr, "Value of potato is %c\n", *potato);
```

...I want to know that the crash was triggered when 0x1234 was dereferenced, as this information may be key in debugging the issue.

We already get this information in our signal handler, so this PR just adds the needed piping to report it to datadog.

# Motivation:

Improve debugging of crashes reported by crashtracker.

# Additional Notes:

I added the `faulting_address` as a tag, similar to `siginfo` and `signame`. We may want to send it as some other thing; no strong opinions there.

# How to test the change?

I was able to test this with the Ruby profiler. I've been trying to update the tests but there's still some failing and I'm kinda struggling with that. Hopefully someone with a bit more rust-fu than me can help out there.

Here's a submitted crash report with a `crashing_address` tag (renamed since this test): https://app.datadoghq.com/dashboard/uwt-n4k-y9a?fromUser=true&refresh_mode=sliding&tpl_var_org_id%5B0%5D=197728&tpl_var_runtime%5B0%5D=ruby&from_ts=1727956296556&to_ts=1727959896556&live=true

### CODE REVIEW

File: bin_tests/tests/crashtracker_bin_test.rs [hunk 1]
```diff
@@ -104,7 +104,8 @@ fn test_crash_tracking_bin(
     assert_eq!(
         serde_json::json!({
           "signum": 11,
-          "signame": "SIGSEGV"
+          "signame": "SIGSEGV",
+          "crash_address": "0x0000000000000000",
```
Comment (r1viollet) [inline bin_tests/tests/crashtracker_bin_test.rs:108]: faulting address ? Trying to find a name that disambiguates from the instruction pointer
I'll add a comment on RFC

Comment (ivoanjo) [reply bin_tests/tests/crashtracker_bin_test.rs:108]: I don't mind renaming it, I think `faulting_address` works nicely as well.

**Review by sanchda — APPROVED** [review]

**Comment by ivoanjo** [comment]
Hmm... there's still two tests that hang and time out in CI:

```
test test_crash_tracking_bin_debug_stdin ... FAILED
test test_crash_tracking_bin_debug_unix_socket ... FAILED
```

I have no idea how to debug these tests :/ . Anyone has any hints on how to find what's failing?

**Comment by ivoanjo** [comment]
Renamed `crash_address` to `faulting_address` in https://github.com/DataDog/libdatadog/pull/663/commits/0d4bb98e732dce251a1d06a1ba5fa35b116fff4a . There's still 2 tests failing, I'm still looking into them.

File: bin_tests/tests/crashtracker_bin_test.rs (hunk 1 above)
Comment (ivoanjo) [reply bin_tests/tests/crashtracker_bin_test.rs]: Renamed `crash_address` to `faulting_address` in https://github.com/DataDog/libdatadog/pull/663/commits/0d4bb98e732dce251a1d06a1ba5fa35b116fff4a

**Comment by ivoanjo** [comment]
Unfortunately I wasn't able yet to figure out what's up with the failing tests... I've ping'd @danielsn to see if he has any idea on how to improve there.

**Comment by sanchda** [comment]
Yeah, looks like an issue on amd64 alpine

**Comment by nsrip-dd** [comment]
> Unfortunately I wasn't able yet to figure out what's up with the failing tests... I've ping'd @danielsn to see if he has any idea on how to fix those.

FWIW I tried those tests on an x86_64 Alpine image on the commit this PR is based on (67d4f7a) and `main` (a76c5ad as of this writing) and they still fail. So... I don't know _why_ they fail, but I don't think this PR breaks them?

**Comment by ivoanjo** [comment]
> FWIW I tried those tests locally on an x86_64 Alpine image on the commit this PR is based on (https://github.com/DataDog/libdatadog/commit/67d4f7a19afd7f7301423e400008d42cc9284c04) and main (https://github.com/DataDog/libdatadog/commit/a76c5ade40c8ac83418fc7ed6bb2f319a874b327 as of this writing) and they still fail. If I comment out [these lines](https://github.com/DataDog/libdatadog/blob/6e2d784b761748a06232a0a1f32cd17fc9355427/crashtracker/src/collector/emitters.rs#L126-L128) they pass. Strangely though, the tests seem to be passing on main and a https://github.com/DataDog/libdatadog/pull/680. So... I don't know why they fail, but I don't think this PR breaks them?

Thanks for giving it a stab! I wonder if the issue is still related to some of the test assertions, but I've found these tests are a bit weird... when they fail they don't give a lot of info; some of the fixes in this PR were kinda trial-and-error. :eyes: 

(I've also merged main into this PR, to make sure it doesn't fall behind)

**Comment by ivoanjo** [comment]
I'm... still stuck on this one. I suspect it may be a real issue; specifically, I see the issue going away if I replace

```rust
        if !sig_info.is_null() && (signum == libc::SIGSEGV || signum == libc::SIGBUS) {
            unsafe { Some((*sig_info).si_addr() as usize) }
        }
```

with something else that does not touch the sig_info. I suspect it may have to do something with chaining the signal handlers. :(

**Review by danielsn — APPROVED** [review]
Approved, with a couple ⚒️

File: crashtracker/src/crash_info/telemetry.rs [hunk 2]
```diff
@@ -191,6 +191,9 @@ fn extract_crash_info_tags(crash_info: &CrashInfo) -> anyhow::Result<String> {
         if let Some(signame) = &siginfo.signame {
             write!(&mut tags, ",signame:{}", signame)?;
         }
+        if let Some(faulting_address) = &siginfo.faulting_address {
+            write!(&mut tags, ",faulting_address:{:#018x}", faulting_address)?;
```
[the remark this answers is not in the document]
Comment (ivoanjo) [reply crashtracker/src/crash_info/telemetry.rs:195]: The two extra chars are for the `0x` prefix :)

File: crashtracker/src/collector/emitters.rs [hunk 3]
```diff
@@ -173,7 +178,17 @@ fn emit_siginfo(w: &mut impl Write, signum: i32) -> anyhow::Result<()> {
     };
 
     writeln!(w, "{DD_CRASHTRACK_BEGIN_SIGINFO}")?;
-    writeln!(w, "{{\"signum\": {signum}, \"signame\": \"{signame}\"}}")?;
+    match faulting_address {
```
[the remark this answers is not in the document]
Comment (ivoanjo) [reply crashtracker/src/collector/emitters.rs]: Fixed in c5c6130a85ec7d686a07f43d92429af72dfa9ee9

**Pull request merged**